File: Android/doraemonkit/src/main/java/com/didichuxing/doraemonkit/kit/logInfo/LogInfoManager.java
package com.didichuxing.doraemonkit.kit.logInfo;

import android.os.Handler;
import android.os.Looper;
import android.os.Message;

import com.didichuxing.doraemonkit.kit.logInfo.reader.LogcatReader;
import com.didichuxing.doraemonkit.kit.logInfo.reader.LogcatReaderLoader;
import com.didichuxing.doraemonkit.util.ExecutorUtil;
import com.didichuxing.doraemonkit.util.LogHelper;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;

/**
 * Created by wanglikun on 2018/10/10.
 */

public class LogInfoManager {
    private static final String TAG = "LogInfoManager";
    private static final int MESSAGE_PUBLISH_LOG = 1001;

    private OnLogCatchListener mListener;

    private LogCatchRunnable mTask;

    private static class Holder {
        private static LogInfoManager INSTANCE = new LogInfoManager();
    }

    private LogInfoManager() {
    }

    public static LogInfoManager getInstance() {
        return Holder.INSTANCE;
    }

    public void start() {
        if (mTask != null) {
            mTask.stop();
        }
        mTask = new LogCatchRunnable();
        ExecutorUtil.execute(mTask);
    }

    public void stop() {
        if (mTask != null) {
            mTask.stop();
        }
    }

    public interface OnLogCatchListener {
        void onLogCatch(List<LogLine> logLine);
    }

    public void registerListener(OnLogCatchListener listener) {
        mListener = listener;
    }

    public void removeListener() {
        mListener = null;
    }

    private static class InternalHandler extends Handler {
        public InternalHandler(Looper looper) {
            super(looper);
        }

        @Override
        public void handleMessage(Message msg) {
            switch (msg.what) {
                case MESSAGE_PUBLISH_LOG: {
                    if (LogInfoManager.getInstance().mListener != null) {
                        LogInfoManager.getInstance().mListener.onLogCatch((List<LogLine>) msg.obj);
                    }
                }
                break;
                default:
                    break;
            }
        }
    }


    private static class LogCatchRunnable implements Runnable {
        private boolean isRunning = true;
        private Handler internalHandler;
        private LogcatReader mReader;
        private int mPid;

        private LogCatchRunnable() {
            internalHandler = new InternalHandler(Looper.getMainLooper());
            mPid = android.os.Process.myPid();
        }

        @Override
        public void run() {
            try {
                LogcatReaderLoader loader = LogcatReaderLoader.create(true);
                mReader = loader.loadReader();

                String line;
                int maxLines = 10000;
                LinkedList<LogLine> initialLines = new LinkedList<>();
                while ((line = mReader.readLine()) != null && isRunning) {
                    LogLine logLine = LogLine.newLogLine(line, false);
                    if (!mReader.readyToRecord()) {
                        if (logLine.getProcessId() == mPid) {
                            initialLines.add(logLine);
                        }
                        if (initialLines.size() > maxLines) {
                            initialLines.removeFirst();
                        }
                    } else if (!initialLines.isEmpty()) {
                        if (logLine.getProcessId() == mPid) {
                            initialLines.add(logLine);
                        }
                        Message message = Message.obtain();
                        message.what = MESSAGE_PUBLISH_LOG;
                        message.obj = new ArrayList<>(initialLines);
                        internalHandler.sendMessage(message);
                        initialLines.clear();
                    } else {
                        // just proceed as normal
                        if (logLine.getProcessId() == mPid) {
                            Message message = Message.obtain();
                            message.what = MESSAGE_PUBLISH_LOG;
                            message.obj = Collections.singletonList(logLine);
                            internalHandler.sendMessage(message);
                        }
                    }
                }
                mReader.killQuietly();
            } catch (IOException e) {
                LogHelper.e(TAG, e.toString());
            }
        }

        public void stop() {
            isRunning = false;
        }
    }
}

File: Android/kotlin/doraemonkit/src/main/java/com/didichuxing/doraemonkit/kit/loginfo/LogInfoManager.kt
package com.didichuxing.doraemonkit.kit.loginfo

import android.os.Handler
import android.os.Looper
import android.os.Message
import com.didichuxing.doraemonkit.kit.loginfo.reader.LogcatReaderLoader
import com.didichuxing.doraemonkit.util.ExecutorUtil
import com.didichuxing.doraemonkit.util.LogHelper
import java.util.*
import java.util.concurrent.atomic.AtomicBoolean
import kotlin.collections.ArrayList

/**
 * @author lostjobs created on 2020/6/28
 */
object LogInfoManager {
    private const val TAG = "LogInfoManager"
    private const val MESSAGE_PUBLISH_LOG = 1001

    private var mListener: OnLogCatchListener? = null
    private var mLogCatchTask: LogCatchRunnable? = null
    fun start() {
        stop()
        mLogCatchTask = LogCatchRunnable().also {
            ExecutorUtil.execute(it)
        }
    }

    fun stop() {
        mLogCatchTask?.stop()
        mLogCatchTask = null
    }

    fun registerListener(listener: OnLogCatchListener) {
        this.mListener = listener
    }

    fun removeListener() {
        this.mListener = null
    }


    interface OnLogCatchListener {
        /**
         * 日志回调
         */
        fun onLogCatch(logLines: List<LogLine>)
    }


    private class InternalHandler(looper: Looper) : Handler(looper) {
        override fun handleMessage(msg: Message) {
            if (msg.what == MESSAGE_PUBLISH_LOG) {
                val logLines = msg.obj
                (logLines as? List<*>)?.filterIsInstance(LogLine::class.java)?.also {
                    mListener?.onLogCatch(it)
                }
            }
        }
    }

    class LogCatchRunnable : Runnable {
        private var isRunning: AtomicBoolean = AtomicBoolean(true)
        private val internalHandler = InternalHandler(Looper.getMainLooper())
        private val mPid = android.os.Process.myPid()

        override fun run() {
            try {
                val loader = LogcatReaderLoader.create(true)
                val maxLines = 10000
                val reader = loader.loadReader()
                val initialLines = LinkedList<LogLine>()
                while (isRunning.get()) {
                    val line = reader.readLine() ?: break
                    val logLine = LogLine.newLogLine(line, false)
                    if (!reader.readyToRecord()) {
                        if (logLine.processId == mPid) {
                            initialLines.add(logLine)
                        }

                        if (initialLines.size > maxLines) {
                            initialLines.removeFirst()
                        }
                    } else if (!initialLines.isEmpty()) {
                        if (logLine.processId == mPid) {
                            initialLines.add(logLine)
                        }

                        val msg = internalHandler.obtainMessage(MESSAGE_PUBLISH_LOG)
                        msg.obj = ArrayList(initialLines)
                        msg.sendToTarget()
                        initialLines.clear()
                    } else {
                        if (logLine.processId == mPid) {
                            val msg = internalHandler.obtainMessage(MESSAGE_PUBLISH_LOG)
                            msg.obj = Collections.singletonList(logLine)
                            msg.sendToTarget()
                        }
                    }
                }
                reader.killQuietly()
            } catch (e: Exception) {
                LogHelper.e(TAG, e.toString())
            }
        }

        fun stop() {
            isRunning.set(false)
        }
    }
}